### DISCUSSION TITLE
PreparedStatement not getting executed

### DISCUSSION THREAD

**Discussion by articlaus** [ing-bank/cassandra-jdbc-wrapper#72] [Q&A]
Hi, jdbc insertion is failing with below error, i am not sure what seems to be causing it

```
java.sql.SQLTransientException: java.lang.IllegalArgumentException: com.datastax.oss.driver.api.core.cql.SimpleStatement referenced from a method is not visible from class loader: 'app'
	at com.ing.data.cassandra.jdbc.CassandraPreparedStatement.<init>(CassandraPreparedStatement.java:156) ~[cassandra-jdbc-wrapper-4.13.0.jar!/:?]
	at com.ing.data.cassandra.jdbc.CassandraConnection.prepareStatement(CassandraConnection.java:579) ~[cassandra-jdbc-wrapper-4.13.0.jar!/:?]
	at com.ing.data.cassandra.jdbc.CassandraConnection.prepareStatement(CassandraConnection.java:557) ~[cassandra-jdbc-wrapper-4.13.0.jar!/:?]

Caused by: java.lang.IllegalArgumentException: com.datastax.oss.driver.api.core.cql.SimpleStatement referenced from a method is not visible from class loader: 'app'
```

**Comment by maximevw** [comment]
Hello @articlaus 

Could you please provide more details about the context of execution (Java version, used framework if any, the code executing the call to the JDBC API, ...)?

**Reply by articlaus** [reply]
> Hello @articlaus
> 
> Could you please provide more details about the context of execution (Java version, used framework if any, the code executing the call to the JDBC API, ...)?

@maximevw 

Sorry for the incomplete questions  below is my imports. Mainly its Spring booth with Reactive Cassandra and Datastax drivers
and java Version is Zulu JDK 11

`    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.7.18</version>
        <relativePath/>
    </parent>
    
            <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-cassandra-reactive</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>ch.qos.logback</groupId>
                    <artifactId>logback-classic</artifactId>
                </exclusion>
                <exclusion>
                    <artifactId>java-driver-core</artifactId>
                    <groupId>com.datastax.oss</groupId>
                </exclusion>
            </exclusions>
        </dependency>
        
               <dependency>
            <groupId>com.ing.data</groupId>
            <artifactId>cassandra-jdbc-wrapper</artifactId>
            <version>4.13.0</version>
        </dependency>
    `

Below is the code snipped using the wrapper for batch inserts

`   

public void updateStatus(DataDTO status) {

        try (CassandraConnection connection = new CassandraConnection(cqlSession, keyspace, ConsistencyLevel.ALL, false, null, null)) {

            connection.setAutoCommit(false);

            final CassandraPreparedStatement statement = connection.prepareStatement(
                   " INSERT INTO data_status (id, index,trace_id,status,pub_status,source_type,last_change_ts,userid) VALUES (?, ?,?,?,?,?,?,?)
            );"

            statement.setObject(1, status.getPrimaryKey().id());


            statement.setString(8, status.getUserId());
            statement.addBatch();
            statement.executeBatch();
            connection.commit();
        } catch (Exception ex) {
            log.error("Exception Occurred while trying to batch insert for status update into Cassandra", ex);
            throw new CustomException(ex.getMessage());
        }
    }`

**Comment by maximevw** [comment]
Hi @articlaus 

It looks like there is a dependency conflict with the Cassandra Java driver. You can try this:
* exclude `com.datastax.oss:java-driver-query-builder` from the `spring-boot-starter-data-cassandra-reactive` depdendency
* downgrade `cassandra-jdbc-wrapper` to 4.12.0 (since 4.13.0, it uses the depdendency `org.apache.cassandra:java-driver-core` instead of the one from `com.datastax.oss`).

Another possibility is that Spring Boot Data Cassandra Reactive has a custom class loader and some classes are not visible from some parts of your code at runtime (this is an hypothesis, I didn't check if it's really the case). 
If you're using Spring DevTools, you may also encounter this kind of errors.

Also, I don't have the view of the whole project, but it's quite strange to import `spring-boot-starter-data-cassandra-reactive` (which uses the native Java driver for Cassandra) to finally use the classic JDBC API. Are you sure this dependency is required or aren't you mixing different ways to perform queries in your code, which could lead to unexpected behaviours like the one you describe?

By the way, your code could also be simplified:
* Cassandra doesn't support transactions, so, changing the auto-commit setting has no effect, nor does calling `commit()` method on the connection.
* You just have one insertion to perform here, so using batch is not necessary if you don't have a loop in your code adding several statements to the batch. Consequently, you could simply use `executeQuery()`.

```java
public void updateStatus(DataDTO status) {
    try (CassandraConnection connection = new CassandraConnection(cqlSession, keyspace, ConsistencyLevel.ALL, false, null, null)) {
        final CassandraPreparedStatement statement = connection.prepareStatement(
               "INSERT INTO data_status (id,index,trace_id,status,pub_status,source_type,last_change_ts,userid) VALUES (?,?,?,?,?,?,?,?)"
        );

        statement.setObject(1, status.getPrimaryKey().id());
        //...
        statement.setString(8, status.getUserId());
        statement.executeQuery();
    } catch (Exception ex) {
        log.error("Exception Occurred while trying to batch insert for status update into Cassandra", ex);
        throw new CustomException(ex.getMessage());
    }
}
```

**Comment by articlaus** [comment]
@maximevw 
Hi When i downgrade, i see that below exception being thrown

`Caused by: java.lang.NoClassDefFoundError: Could not initialize class com.ing.data.cassandra.jdbc.types.DataTypeEnum\n\tat com.ing.data.cassandra.jdbc.CassandraPreparedStatement.setInt(CassandraPreparedStatement.java:468)`

and i tried adding Wrapper codecs to session building still fails with same exception.

`"SeverityText":"WARN","Body":"[s1] Ignoring codec com.ing.data.cassandra.jdbc.codec.VarintToIntCodec@30099275 because it collides with previously registered codec com.ing.data.cassandra.jdbc.codec.VarintToIntCodec@1e275b26","Resource":{"app.version":"1.0.0-SNAPSHOT","host.id":"192.168.0.225","env":"stage","app":"data-tool",},"Attributes":{"code.namespace":"com.datastax.oss.driver.internal.core.type.codec.registry.CachingCodecRegistry","code.function":"register","code.lineno":"134","thread.id":"606"`

`            sessionBuilder.addTypeCodecs(new TimestampToLongCodec(),
                    new LongToIntCodec(),
                    new IntToLongCodec(),
                    new BigintToBigDecimalCodec(),
                    new DecimalToDoubleCodec(),
                    new FloatToDoubleCodec(),
                    new VarintToIntCodec(),
                    new TinyintToIntCodec(),
                    new SmallintToIntCodec()
            );`

even then the above WARN, gets printed constantly

**Comment by articlaus** [comment]
@maximevw 
I forgot to mention, the reason i am using the wrapper along with Reactive was, i wanted to do batch insert, like millions of rows into cassandra in timely manner. while using the Reactive as a simple selector etc. because reactive cassandra was just too slow. and while going through the documentation, it wasn't clear on how to setup connections, specially if my cassandra has credentials what would be the correct way to create connection if i want to connect to cassandra with credentials and i want to setup custom properties like Timeouts etc? as previously i was using Datastax driver to make connection etc. now i couldn't figure out how to setup just using the Wrapper.

**Reply by maximevw** [reply]
Hi @articlaus 
> When i downgrade, i see that below exception being thrown
> 
> `Caused by: java.lang.NoClassDefFoundError: Could not initialize class com.ing.data.cassandra.jdbc.types.DataTypeEnum\n\tat com.ing.data.cassandra.jdbc.CassandraPreparedStatement.setInt(CassandraPreparedStatement.java:468)`

This looks very much like a classpath issue. `DataTypeEnum` is present in the wrapper for years, so for some reason, at runtime your app doesn't see all the classes of the wrapper.  Maybe running a command like `mvn dependency:tree` could help to identify if different versions of the same dependency are imported or if there are some conflicts.

> and i tried adding Wrapper codecs to session building still fails with same exception.
> 
> `"SeverityText":"WARN","Body":"[s1] Ignoring codec com.ing.data.cassandra.jdbc.codec.VarintToIntCodec@30099275 because it collides with previously registered codec com.ing.data.cassandra.jdbc.codec.VarintToIntCodec@1e275b26","Resource":{"app.version":"1.0.0-SNAPSHOT","host.id":"192.168.0.225","env":"stage","app":"data-tool",},"Attributes":{"code.namespace":"com.datastax.oss.driver.internal.core.type.codec.registry.CachingCodecRegistry","code.function":"register","code.lineno":"134","thread.id":"606"`
> 
> ` sessionBuilder.addTypeCodecs(new TimestampToLongCodec(), new LongToIntCodec(), new IntToLongCodec(), new BigintToBigDecimalCodec(), new DecimalToDoubleCodec(), new FloatToDoubleCodec(), new VarintToIntCodec(), new TinyintToIntCodec(), new SmallintToIntCodec() );`
> 
> even then the above WARN, gets printed constantly

It's not necessary to add these codecs by yourself, it's automatically done when you create a connection re-using a pre-existing session. Adding them generates the warning if the codecs have already been added to the session. But, you found something interesting, the wrapper does not check if the codecs are already present before adding them to the session. So, I noted the bug (not blocking) to fix it in the next version.
 

> I forgot to mention, the reason i am using the wrapper along with Reactive was, i wanted to do batch insert, like millions of rows into cassandra in timely manner. while using the Reactive as a simple selector etc. because reactive cassandra was just too slow. and while going through the documentation, it wasn't clear on how to setup connections, specially if my cassandra has credentials what would be the correct way to create connection if i want to connect to cassandra with credentials and i want to setup custom properties like Timeouts etc? as previously i was using Datastax driver to make connection etc. now i couldn't figure out how to setup just using the Wrapper.

Until now, the wrapper is mainly designed to instantiate connections using a JDBC URL as explained in this page: https://github.com/ing-bank/cassandra-jdbc-wrapper/wiki/JDBC-driver-and-connection-string#basic-usage. JDBC URLs support a lot of parameters to fine-tune the connection (timeouts, SSL, ...).
It's also possible to obtain a connection from a `CassandraDataSource` object. But some properties are not settable for now. In the next version, it will be possible to set all the existing properties of the wrapper using `CassandraDataSource`. For now, you can at least specify some basic information such as the credentials.
